File: pipeline/reimpl/cotrn00c.py
"""
Reimplementation of COTRN00C — CardDemo Transaction List Screen.

CICS online program. Lists transactions (10 per page) from the TRANSACT file,
with PF7/PF8 navigation and selection to view a transaction detail.

Selection: Enter 'S' in a row's selector to view that transaction → COTRN01C
"""

from __future__ import annotations
from dataclasses import dataclass, field
from decimal import Decimal
from typing import Optional

from .carddemo_data import (
    CarddemoCommarea, ScreenHeader, TranRecord,
    CCDA_MSG_INVALID_KEY, DFHENTER, DFHPF3, DFHPF7, DFHPF8,
)


WS_PGMNAME = "COTRN00C"
WS_TRANID  = "CT00"
PAGE_SIZE  = 10


@dataclass
class TranListRow:
    tran_id: str = ""
    tran_type_cd: str = ""
    tran_cat_cd: int = 0
    tran_source: str = ""
    tran_desc: str = ""
    tran_amt: Decimal = Decimal("0.00")
    tran_card_num: str = ""
    tran_proc_ts: str = ""
    sel_flag: str = ""


@dataclass
class TranListResult:
    rows: list[TranListRow] = field(default_factory=list)
    page_num: int = 1
    has_next_page: bool = False
    has_prev_page: bool = False
    xctl_program: str = ""
    selected_tran_id: str = ""
    commarea: Optional[CarddemoCommarea] = None
    screen: Optional[ScreenHeader] = None
    message: str = ""
    error: bool = False
    return_to_prev: bool = False


class TranRepository:
    def __init__(self, transactions: list[TranRecord]):
        self._trans = sorted(transactions, key=lambda t: t.tran_id)

    def get_page(self, page_num: int) -> tuple[list[TranRecord], bool]:
        start = (page_num - 1) * PAGE_SIZE
        end = start + PAGE_SIZE
        return self._trans[start:end], len(self._trans) > end


def process_tran_list(
    eibcalen: int,
    eibaid: str,
    commarea: CarddemoCommarea,
    tran_repo: TranRepository,
    page_num: int = 1,
    selected_rows: list[tuple[str, str]] = None,
) -> TranListResult:
    """Process transaction list screen — mirrors COTRN00C PROCEDURE DIVISION."""
    result = TranListResult()

    if eibcalen == 0:
        result.return_to_prev = True
        result.xctl_program = "COSGN00C"
        return result

    result.commarea = commarea

    if commarea.cdemo_pgm_context == 0:
        commarea.cdemo_pgm_context = 1
        return _load_page(1, tran_repo, commarea, result)

    if eibaid in (DFHENTER, "ENTER"):
        return _process_enter(selected_rows or [], page_num, tran_repo, commarea, result)
    elif eibaid in (DFHPF3, "PF3"):
        back = commarea.cdemo_from_program or "COMEN01C"
        result.xctl_program = back
        result.return_to_prev = True
        result.commarea = commarea
        return result
    elif eibaid in (DFHPF7, "PF7"):
        return _load_page(max(1, page_num - 1), tran_repo, commarea, result)
    elif eibaid in (DFHPF8, "PF8"):
        return _load_page(page_num + 1, tran_repo, commarea, result)
    else:
        result.error = True
        result.message = CCDA_MSG_INVALID_KEY
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.screen.errmsg = result.message
        result.commarea = commarea
        return _load_page(page_num, tran_repo, commarea, result)


def _load_page(page_num, repo, commarea, result):
    trans, has_next = repo.get_page(page_num)
    result.page_num = page_num
    result.has_next_page = has_next
    result.has_prev_page = page_num > 1
    result.rows = [
        TranListRow(
            tran_id=t.tran_id,
            tran_type_cd=t.tran_type_cd,
            tran_cat_cd=t.tran_cat_cd,
            tran_source=t.tran_source,
            tran_desc=t.tran_desc[:40],
            tran_amt=t.tran_amt,
            tran_card_num=t.tran_card_num,
            tran_proc_ts=t.tran_proc_ts[:10],
        )
        for t in trans
    ]
    result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
    result.commarea = commarea
    return result


def _process_enter(selected_rows, page_num, repo, commarea, result):
    for sel_flag, tran_id in selected_rows:
        if sel_flag and sel_flag.strip() not in ("", "\x00") and tran_id.strip():
            result.selected_tran_id = tran_id.strip()
            commarea.cdemo_pgm_context = 0
            result.xctl_program = "COTRN01C"
            result.commarea = commarea
            return result
    return _load_page(page_num, repo, commarea, result)


# ── Differential harness runner adapter ──────────────────────────────────────

_SEED_TRANS = [
    TranRecord(tran_id=f"{i:016d}", tran_type_cd="01", tran_cat_cd=1,
               tran_source="ONLINE", tran_desc=f"PURCHASE {i:04d}",
               tran_amt=Decimal(f"{i * 10}.00"),
               tran_card_num="4111111111111111",
               tran_proc_ts="2025-01-15-10.30.00.000000")
    for i in range(1, 13)
]


def run_vector(inputs: dict) -> dict:
    """Canonical runner entry point for the differential harness.

    SCENARIO selects a hardcoded test path:
      LIST_PAGE_1   — load first page (10 rows)
      LIST_PAGE_2   — load second page (2 rows)
      SELECT_TRAN   — select a transaction from list → COTRN01C
      INVALID_KEY   — press PF9 → invalid key error
      PF3_RETURN    — press PF3 → return to calling program
    """
    scenario = inputs.get("SCENARIO", "LIST_PAGE_1")

    repo = TranRepository(list(_SEED_TRANS))

    commarea = CarddemoCommarea(
        cdemo_from_tranid="CM00",
        cdemo_from_program="COMEN01C",
        cdemo_user_id="USER0001",
        cdemo_user_type="U",
        cdemo_pgm_context=1,
    )

    if scenario == "LIST_PAGE_1":
        result = process_tran_list(
            eibcalen=100, eibaid="ENTER", commarea=commarea,
            tran_repo=repo, page_num=1, selected_rows=[],
        )
    elif scenario == "LIST_PAGE_2":
        result = process_tran_list(
            eibcalen=100, eibaid="PF8", commarea=commarea,
            tran_repo=repo, page_num=1, selected_rows=[],
        )
    elif scenario == "SELECT_TRAN":
        result = process_tran_list(
            eibcalen=100, eibaid="ENTER", commarea=commarea,
            tran_repo=repo, page_num=1,
            selected_rows=[("S", "0000000000000003")],
        )
    elif scenario == "INVALID_KEY":
        result = process_tran_list(
            eibcalen=100, eibaid="PF9", commarea=commarea,
            tran_repo=repo, page_num=1, selected_rows=[],
        )
    elif scenario == "PF3_RETURN":
        result = process_tran_list(
            eibcalen=100, eibaid="PF3", commarea=commarea,
            tran_repo=repo, page_num=1, selected_rows=[],
        )
    else:
        result = process_tran_list(
            eibcalen=100, eibaid="ENTER", commarea=commarea,
            tran_repo=repo, page_num=1, selected_rows=[],
        )

    out: dict[str, str] = {
        "PAGE_NUM": str(result.page_num),
        "ROW_COUNT": str(len(result.rows)),
        "HAS_NEXT": "Y" if result.has_next_page else "N",
        "HAS_PREV": "Y" if result.has_prev_page else "N",
        "ERROR": "Y" if result.error else "N",
        "MESSAGE": result.message,
        "XCTL_PROGRAM": result.xctl_program or "",
        "SELECTED_TRAN_ID": result.selected_tran_id,
    }
    for i, row in enumerate(result.rows):
        out[f"ROW_{i}_TRAN_ID"] = row.tran_id
        out[f"ROW_{i}_AMT"] = f"{row.tran_amt:.2f}"
        out[f"ROW_{i}_DESC"] = row.tran_desc
    return out

File: pipeline/reimpl/java/runner/src/main/java/com/modernization/masquerade/runner/programs/Cotrn00c.java
package com.modernization.masquerade.runner.programs;

import com.modernization.masquerade.runner.ProgramRunner;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Java reimplementation of COTRN00C — CardDemo Transaction List Screen.
 *
 * <p>Mirrors {@code pipeline/reimpl/cotrn00c.py}. Python is the source of truth.
 *
 * <p>Standalone business method: {@link #processTranList}. The {@link #runVector}
 * method is a thin adapter that builds scenario-specific parameters and delegates.
 */
public class Cotrn00c implements ProgramRunner {

    private static final int PAGE_SIZE = 10;
    private static final String CCDA_MSG_INVALID_KEY = "Invalid key pressed.                    ";

    // ── Inner data types ─────────────────────────────────────────────────

    static class TranRecord {
        final String tranId;
        final String tranTypeCd;
        final String tranDesc;
        final BigDecimal tranAmt;
        final String tranCardNum;
        final String tranProcTs;

        TranRecord(String tranId, String tranTypeCd, String tranDesc,
                   BigDecimal tranAmt, String tranCardNum, String tranProcTs) {
            this.tranId = tranId;
            this.tranTypeCd = tranTypeCd;
            this.tranDesc = tranDesc;
            this.tranAmt = tranAmt;
            this.tranCardNum = tranCardNum;
            this.tranProcTs = tranProcTs;
        }
    }

    static class TranListRow {
        final String tranId;
        final String tranDesc;
        final BigDecimal tranAmt;

        TranListRow(String tranId, String tranDesc, BigDecimal tranAmt) {
            this.tranId = tranId;
            this.tranDesc = tranDesc;
            this.tranAmt = tranAmt;
        }
    }

    static class TranListResult {
        List<TranListRow> rows = new ArrayList<>();
        int pageNum = 1;
        boolean hasNext = false;
        boolean hasPrev = false;
        boolean error = false;
        String message = "";
        String xctlProgram = "";
        String selectedTranId = "";
    }

    static class TranRepo {
        private final List<TranRecord> trans;

        TranRepo(List<TranRecord> trans) {
            this.trans = trans;
            this.trans.sort((a, b) -> a.tranId.compareTo(b.tranId));
        }

        /** Return page rows and whether there is a next page. */
        PageResult getPage(int pageNum) {
            int start = (pageNum - 1) * PAGE_SIZE;
            int end = Math.min(start + PAGE_SIZE, trans.size());
            List<TranRecord> page = (start < trans.size())
                    ? trans.subList(start, end) : new ArrayList<>();
            return new PageResult(page, trans.size() > end);
        }
    }

    static class PageResult {
        final List<TranRecord> records;
        final boolean hasNext;
        PageResult(List<TranRecord> records, boolean hasNext) {
            this.records = records;
            this.hasNext = hasNext;
        }
    }

    // ── Standalone business method ──────────────────────────────────────

    /**
     * Process transaction list screen — mirrors Python's process_tran_list.
     *
     * @param eibcalen     CICS EIBCALEN
     * @param eibaid       aid key pressed
     * @param pgmContext   commarea program context (0 = initial)
     * @param fromProgram  commarea cdemo_from_program
     * @param tranRepo     transaction repository
     * @param pageNum      current page number
     * @param selectedRows list of (selFlag, tranId) pairs from screen
     * @return TranListResult with all output fields
     */
    TranListResult processTranList(
            int eibcalen,
            String eibaid,
            int pgmContext,
            String fromProgram,
            TranRepo tranRepo,
            int pageNum,
            List<String[]> selectedRows
    ) {
        TranListResult result = new TranListResult();

        if (eibcalen == 0) {
            result.xctlProgram = "COSGN00C";
            return result;
        }

        if (pgmContext == 0) {
            return loadPage(1, tranRepo, result);
        }

        switch (eibaid) {
            case "ENTER":
                return processEnter(selectedRows, pageNum, tranRepo, result);
            case "PF3": {
                String back = (fromProgram != null && !fromProgram.isEmpty()) ? fromProgram : "COMEN01C";
                result.xctlProgram = back;
                return result;
            }
            case "PF7":
                return loadPage(Math.max(1, pageNum - 1), tranRepo, result);
            case "PF8":
                return loadPage(pageNum + 1, tranRepo, result);
            default:
                result.error = true;
                result.message = CCDA_MSG_INVALID_KEY;
                return loadPage(pageNum, tranRepo, result);
        }
    }

    private TranListResult loadPage(int pageNum, TranRepo repo, TranListResult result) {
        PageResult pr = repo.getPage(pageNum);
        result.pageNum = pageNum;
        result.hasNext = pr.hasNext;
        result.hasPrev = pageNum > 1;
        result.rows = new ArrayList<>();
        for (TranRecord t : pr.records) {
            String desc = t.tranDesc.length() > 40 ? t.tranDesc.substring(0, 40) : t.tranDesc;
            result.rows.add(new TranListRow(t.tranId, desc, t.tranAmt));
        }
        return result;
    }

    private TranListResult processEnter(
            List<String[]> selectedRows, int pageNum,
            TranRepo repo, TranListResult result
    ) {
        if (selectedRows != null) {
            for (String[] sel : selectedRows) {
                String selFlag = sel[0];
                String tranId = sel[1];
                if (selFlag != null && !selFlag.trim().isEmpty()
                        && !selFlag.trim().equals("\0")
                        && tranId != null && !tranId.trim().isEmpty()) {
                    result.selectedTranId = tranId.trim();
                    result.xctlProgram = "COTRN01C";
                    return result;
                }
            }
        }
        return loadPage(pageNum, repo, result);
    }

    // ── Seed data ───────────────────────────────────────────────────────

    private static final int SEED_COUNT = 12;

    private List<TranRecord> buildSeedTrans() {
        List<TranRecord> list = new ArrayList<>();
        for (int i = 1; i <= SEED_COUNT; i++) {
            list.add(new TranRecord(
                    String.format("%016d", i),
                    "01",
                    String.format("PURCHASE %04d", i),
                    new BigDecimal(i * 10),
                    "4111111111111111",
                    "2025-01-15-10.30.00.000000"
            ));
        }
        return list;
    }

    // ── Thin runVector adapter ──────────────────────────────────────────

    @Override
    public Map<String, String> runVector(Map<String, String> inputs) {
        String scenario = inputs.getOrDefault("SCENARIO", "LIST_PAGE_1");

        TranRepo repo = new TranRepo(buildSeedTrans());
        List<String[]> noSelection = new ArrayList<>();

        TranListResult result;

        switch (scenario) {
            case "LIST_PAGE_1":
                result = processTranList(100, "ENTER", 1, "COMEN01C",
                        repo, 1, noSelection);
                break;

            case "LIST_PAGE_2":
                result = processTranList(100, "PF8", 1, "COMEN01C",
                        repo, 1, noSelection);
                break;

            case "SELECT_TRAN": {
                List<String[]> sel = new ArrayList<>();
                sel.add(new String[]{"S", "0000000000000003"});
                result = processTranList(100, "ENTER", 1, "COMEN01C",
                        repo, 1, sel);
                break;
            }

            case "INVALID_KEY":
                result = processTranList(100, "PF9", 1, "COMEN01C",
                        repo, 1, noSelection);
                break;

            case "PF3_RETURN":
                result = processTranList(100, "PF3", 1, "COMEN01C",
                        repo, 1, noSelection);
                break;

            default:
                result = processTranList(100, "ENTER", 1, "COMEN01C",
                        repo, 1, noSelection);
                break;
        }

        Map<String, String> out = new LinkedHashMap<>();
        out.put("PAGE_NUM", String.valueOf(result.pageNum));
        out.put("ROW_COUNT", String.valueOf(result.rows.size()));
        out.put("HAS_NEXT", result.hasNext ? "Y" : "N");
        out.put("HAS_PREV", result.hasPrev ? "Y" : "N");
        out.put("ERROR", result.error ? "Y" : "N");
        out.put("MESSAGE", result.message);
        out.put("XCTL_PROGRAM", result.xctlProgram);
        out.put("SELECTED_TRAN_ID", result.selectedTranId);

        for (int i = 0; i < result.rows.size(); i++) {
            TranListRow row = result.rows.get(i);
            out.put("ROW_" + i + "_TRAN_ID", row.tranId);
            out.put("ROW_" + i + "_AMT", String.format("%.2f", row.tranAmt));
            out.put("ROW_" + i + "_DESC", row.tranDesc);
        }

        return out;
    }
}
